controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_types.go

```diff
@@ -158,6 +181,40 @@ type RollingUpdate struct {
 	MaxSurge *intstr.IntOrString `json:"maxSurge,omitempty"`
 }
 
+// RemediationStrategy allows to define how control plane machine remediation happens.
+type RemediationStrategy struct {
+	// MaxRetry is the Max number of retries while attempting to remediate an unhealthy machine.
+	// A retry happens when a machine that was created as a replacement for an unhealthy machine also fails.
+	// For example, given a control plane with three machines M1, M2, M3:
+	//
+	//	M1 become unhealthy; remediation happens, and M1-1 is created as a replacement.
+	//	If M1-1 (replacement of M1) has problems while bootstrapping it will become unhealthy, and then be
+	//	remediated; such operation is considered a retry, remediation-retry #1.
+	//	If M1-2 (replacement of M1-2) becomes unhealthy, remediation-retry #2 will happen, etc.
+	//
+	// A retry could happen only after RetryPeriod from the previous retry.
+	// If a machine is marked as unhealthy after MinHealthyPeriod from the previous remediation expired,
+	// this is not considered a retry anymore because the new issue is assumed unrelated from the previous one.
+	//
+	// If not set, the remedation will be retried infinitely.
+	// +optional
+	MaxRetry *int32 `json:"maxRetry,omitempty"`
```

Comment:
Same here, could we set a default upper limit that we think it's reasonable?

Reply:
I think it's a valid use case to want infinite retries. What if you don't care about the number of machine remediations but want everything to recover automatically without manual intervention when the infra provider recovers?

Reply:
Sorry did you ask for a default or an upper limit or both?

Reply:
Also in this case, If we add a defaulting rule, then it will be impossible for the users to ask for "retry indefinitely", which I consider a valid use case.

Also, retrying indefinitely is the current behavior and AFAIK no one complained about it, so I do not think it is necessary to force the system to always stop retrying at a certain point.

What we can do in a follow-up iteration is to add a sort of exponential back-off between retries, but this needs some more thinking on how to track intervals between iterations and if/how this should interact with MinHealthyPeriod.

Reply:
Despite what is technically feasible, let's use current behavior as a default since we don't have demand or a set of trusted data to use for picking up other values.